### ISSUE TITLE
Model parameters averaging TF and Keras

### ISSUE THREAD

**Issue by SoulGuedria** [horovod/horovod#1053]
**Environment:**
1. Framework: TensorFlow, Keras
2. Horovod version: 0.15.2
4. MPI version:(OpenRTE) 3.0.0
5. CUDA version: release 9.0, V9.0.176
6. Python version: python3
8. OS and version: debian 9

**My question:**

Hello,
I have two questions about the model parameters averaging during the training process using Horvod with TF and Keras:
 
1) Is it mandatory to add the hvd.callbacks.MetricAverageCallback to the list of callbacks in order to do the parameters averaging at the end of every epoch ?

2) How to modify the gradient averaging synchronization period between all horovod workors with TF and Keras from one epoch to X mini-batchs?

Thanks a lot :)

**Comment by alsrgv** [comment]
@SoulGuedria, sorry for the delayed response.

1. `MetricAverageCallback` averages metrics at the end of every epoch - this helps to accelerate validation since every worker can run validation on a part of the dataset and then metrics can be averaged.  Actual gradient averaging is done by `DistributedOptimizer`.

2. Gradient averaging actually happens every mini-batch.  Are you asking how to make it happen less frequently?

**Comment by SoulGuedria** [comment]
@alsrgv Thanks for your reply and for this nice project. 

1. If I understood correctly, the gradient averaging happens every mini-batch including the end of every epoch. 

2. I might have a stupid question. Do metrics averaging and gradient averaging stand for the same computational process ? If not, what do you mean by metrics averaging ?

3. I'm asking these questions because i'm evaluating a 18 GPUs horovod-based CNN for semantic segmentation. I noticed an accuracy loss of 2% for distributed horovod comparing to a single GPU even though i'm using the Facebook linear scaling rule for the learning rate. Do you have a possible explanation for this accuracy loss and how could I avoid it ? Thanks a lot :)

**Comment by alsrgv** [comment]
1. That's right.
2. Metrics are things like training accuracy or validation accuracy.
3. Are you doing learning rate warmup, [like this](https://github.com/horovod/horovod/blob/master/examples/keras_imagenet_resnet50.py#L147)?

**Comment by SoulGuedria** [comment]
Yes exactly i'm doing gradual learning rate warmup for the first 5 epochs.

**Comment by alsrgv** [comment]
@SoulGuedria, interesting - may be a good idea to play around with hyperparameters then. Which optimizer are you using?

**Comment by SoulGuedria** [comment]
@alsrgv, I'm using Adam optimizer. I tried also SGD. Yes playing around with hyperparameters sounds a good idea. I will try other combinations of hyperparameters and i will keep you informed. Thanks for the suggestion :)
